feat(lidar_transfusion): add lidar_transfusion 3D detection package

amadeuszsz opened the pull request:
## Description

<!-- Write a brief description of this PR. -->

The PR adds a new `lidar_transfusion` package which brings a new 3D detection component based on TransFusion[1].

https://github.com/autowarefoundation/autoware.universe/assets/37396312/0c2e3f64-b72f-4cd6-b72b-fa50525ea9b8

## Related links

<!-- Write the links related to this PR. Private links should be clearly marked as private, for example, '[FOO COMPANY INTERNAL LINK](https://example.com)'. -->
[1] [Xuyang Bai, Zeyu Hu, Xinge Zhu, Qingqiu Huang, Yilun Chen, Hongbo Fu and Chiew-Lan Tai. "TransFusion: Robust LiDAR-Camera Fusion for 3D Object Detection with Transformers." arXiv preprint arXiv:2203.11496 (2022)](https://arxiv.org/abs/2203.11496)
[2] [rosbag (TIER IV internal link)](https://drive.google.com/file/d/1oQ3gn84q_lx2BrHYJ7usVMmtHn-VHagK/view?usp=sharing)
[3] [helper files (TIER IV internal link)](https://drive.google.com/drive/folders/1IQeD6NsLqjUYbuJYhqgWsO9IsRNY9EWy?usp=sharing)
[4] [TransFusion onnx model (TIER IV internal link)](https://drive.google.com/file/d/13HOLSpSeEOFAF-pZ0szWAQy7VyvuA0QW/view?usp=sharing)

## Tests performed
* Model metrics:
```
------------- T4Metric results -------------
04/25 09:28:12 - mmengine - INFO - | class_name | mAP   | AP@0.5m | AP@1.0m | AP@2.0m | AP@4.0m | error@trans_err | error@scale_err | error@orient_err | error@vel_err | error@attr_err |
04/25 09:28:12 - mmengine - INFO - |----------------------------------------------------------------------------------------------------------------------------------------------------|
04/25 09:28:12 - mmengine - INFO - | car        | 0.788 | 0.609   | 0.798   | 0.859   | 0.884   | 0.26            | 0.159           | 0.103            | 0.779         | 1              |
04/25 09:28:12 - mmengine - INFO - | truck      | 0.59  | 0.315   | 0.577   | 0.7     | 0.766   | 0.379           | 0.149           | 0.0516           | 0.77          | 1              |
04/25 09:28:12 - mmengine - INFO - | bus        | 0.554 | 0.347   | 0.532   | 0.628   | 0.709   | 0.363           | 0.137           | 0.0762           | 1.69          | 1              |
04/25 09:28:12 - mmengine - INFO - | bicycle    | 0.236 | 0.192   | 0.241   | 0.251   | 0.261   | 0.283           | 0.248           | 0.27             | 0.839         | 1              |
04/25 09:28:12 - mmengine - INFO - | pedestrian | 0.628 | 0.57    | 0.613   | 0.642   | 0.688   | 0.253           | 0.281           | 0.424            | 0.626         | 1              |
04/25 09:28:12 - mmengine - INFO - | Total mAP: 0.559
```
* ROS 2 node performance  with RTX 4090:

```
                   | pre-process [ms] | inference [ms] | post-process [ms] | total [ms]  |
------------------------------------------------------------------------------------------
lidar_transfusion  |   1.52 ± 0.14    |  2.29 ± 0.37   |    0.20 ± 0.15    | 4.07 ± 0.44 |
lidar_centerpoint  |        -         |       -        |         -         | 5.53 ± 0.55 |
```

<!-- Describe how you have tested this PR. -->

## Notes for reviewers

<!-- Write additional information if necessary. It should be written if there are related PRs that should be merged at the same time. -->
The package can best tested with rosbag file. If data needed, you can use this rosbag[2] and copy helper files[3] to the `launch` directory of `lidar_transfusion` package before building. The default path for `onnx` model is `~/autoware_data/lidar_transfusion/transfusion.onnx`. The model awaits for deployment, temporary please use attached link[4].

To start, run commands:
```bash
# terminal 1
ros2 launch lidar_transfusion data_pipeline.launch.py 

# terminal 2
ros2 bag play trailer_yaw_ticket_data_jpntaxi7/01d060d9-8d25-45e0-b45e-9fd7201ac27b_2023-05-26-11-30-03_p0900_3.db3 --loop --clock

# terminal 3
rviz2 --ros-args -p use_sim_time:=True 

# terminal 4 (for convenience you can add <param name="use_sim_time" value="true"/> to the node in xml file)
ros2 launch lidar_transfusion lidar_transfusion.launch.xml
```

## Interface changes

<!-- Describe any changed interfaces, such as topics, services, or parameters. -->

## Effects on system behavior

<!-- Describe how this PR affects the system behavior. -->

## Pre-review checklist for the PR author

The PR author **must** check the checkboxes below when creating the PR.

- [x] I've confirmed the [contribution guidelines].
- [x] The PR follows the [pull request guidelines].

## In-review checklist for the PR reviewers

The PR reviewers **must** check the checkboxes below before approval.

- [ ] The PR follows the [pull request guidelines].
- [ ] The PR has been properly tested.
- [ ] The PR has been reviewed by the code owners.

## Post-review checklist for the PR author

The PR author **must** check the checkboxes below before merging.

- [x] There are no open discussions or they are tracked via tickets.
- [x] The PR is ready for merge.

After all checkboxes are checked, anyone who has write access can merge the PR.

[contribution guidelines]: https://autowarefoundation.github.io/autoware-documentation/main/contributing/
[pull request guidelines]: https://autowarefoundation.github.io/autoware-documentation/main/contributing/pull-request-guidelines/

knzo25 commented:
Comments:
 - When the number of voxels falls outside the expected parameters, the output is 200 NaN objects (200 being the max number of objects). This can happen when the top lidar is missing in the taxi, for example. We should consider lowering the value, and/or aborting the inference + adding an error message.
 - We need to add transfusion to the perception related launchers. I did it in my local environment, so I will either pass them to the PR's author, commit them myself, or explain him the procedure.
 - In my environment, I have cuda memory errors sporadically. It happens inside the first kernel of the preprocess step.
 - Not confirmed, but the stream used in the preprocessing may not be the one we are expecting (since adding device synchronizations instead of streams changed the behavior)

knzo25 commented:
@taikitanaka3 
The PR can still be reviewed to some degree, but I think it is better to wait until the errors disappear :pray:

taikitanaka3 commented:
@knzo25 
I think you wanted to mention @scepter914

knzo25 commented:
Apologies :bow:

amadeuszsz commented:
Thank you @knzo25 for your review. Let me investigate the kernels first to find mentioned leak. Regarding the NaN output, as you suggested we will:
- change the optimization profile,
- handle the cases with not sufficient number of voxels after preprocessing for optimization profile.

amadeuszsz commented:
@knzo25 
Recent fix solves the issue with cuda memory and NaNs. Please check if the issue disappeared on your machine as well.

scepter914 reviewed (commented):
Thank you for PR.
Overall very readable code and rich documentation is so great. :+1: 

I took a quick look at whole code and I ask you to add some members for maintainers because development will be robuster and smoother if multiple people can maintain it.
As the rest of my work, I'll test with some rosbag and I'll approve after I confirm the operation with some Rosbag.

scepter914 commented:
@amadeuszsz 

I apologize for the inconvenience, would you fix [DCO](https://github.com/autowarefoundation/autoware.universe/pull/6890/checks?check_run_id=24713916803)?
We need to pass CI include DCO for merge (this specification is so invonvenient...).

scepter914 reviewed (approved):
I confirm the perform with other rosbags.
After kenzo-san's review and uploading model, we can merge this PR. :+1+

knzo25 commented:
@amadeuszsz 
I added transfusion to the other universe launchers.
Just for the duration of the review, I set transfusion as the default model. However, before merging we should return it to centerpoint, since although working, there has not been a proper evaluation in all the respective places

scepter914 commented:
@amadeuszsz 
Now web team uploads the model and the param files !

https://awf.ml.dev.web.auto/perception/models/transfusion/v1/transfusion.onnx
https://awf.ml.dev.web.auto/perception/models/transfusion/v1/lidar_transfusion.param.yaml
https://awf.ml.dev.web.auto/perception/models/transfusion/v1/detection_class_remapper.param.yaml

amadeuszsz commented:
@scepter914 
To keep the consistency we will rename config deployment `v1/lidar_transfusion.param.yaml` -> `v1/transfusion.param.yaml`.  Please wait for the deployment.

knzo25 commented:
@amadeuszsz 
While doing other task that was asked of me, I was looking at the whole perception task and saw the preprocessing being done in cpu in centerpoint. I kind of forgot you told me you were doing it already in gpu for transfusion so I ended up doing double work and implemented it myself there as well :bow: 

When I sent the PR I remembered, and after comparing the implementation, while quite similar, they have different overall costs, so I wonder if you can change the implementation here so it has around the same number of operations:

Let `K` be the total number of frames used for inference and `N` the number of points in a pointcloud.
Assuming that the queue if full, the preprocessing before proper voxelization cost would be as follows:

transfusion:
enqueue: 
 - N* copy(host->host)
transform and concat:
 - move to gpu: K*N*copy(host->device)
 - kernel: K*N*transform(device->device)

(proposed) centerpoint:
enqueue:
 - N * copy (host->device)
transform and concat:
 - kernel: K*N*kernel(device->device)

knzo25 commented:
@amadeuszsz 
After reading your kernels seriously this time, I realized that you went to great lengths to make the operations general.
Usually, when writing kernels (cpu or gpu), you want to avoid all the not strictly needed branches and loops.
Could you specialize the kernels so that the fastest implementation is used on runtime? (maybe leaving your generic version to the case that no fast implementation is available)

Also, since we control the driver, you usually would not need to pay much attention to reverting the endianness

knzo25 commented on perception/lidar_transfusion/include/lidar_transfusion/preprocess/preprocess_kernel.hpp:
```diff
@@ -0,0 +1,72 @@
+
+  cudaError_t generateVoxelsInput_launch(
+    uint8_t * cloud_data, CloudInfo & cloud_info, unsigned int points_agg, unsigned int points_size,
+    float time_lag, float * affine_past2current, float * points);
```
In this context, the transformation does not have the meaning of `affine_past2current` so it makes more sense to have a generic name

amadeuszsz replied:
In this context the `affine_past2current` exactly mean affine transformation matrix. Could you check it again?

knzo25 replied:
I meant that while in 
https://github.com/amadeuszsz/autoware.universe/blob/63a30098d1c9f6e71e50e96ab0d46acb811a3c05/perception/lidar_transfusion/lib/preprocess/voxel_generator.cpp#L68
The variables has the meaning of a transformation between different time stamps,
in the kernel that nuance is lot and you have kernel that applies a transformation, not necessarily one between two time points. As such, I would recommend a name like `affine_transform`

amadeuszsz replied:
https://github.com/autowarefoundation/autoware.universe/pull/6890/commits/3c0187f3d1bc8aacc640ea313016cff8b2e7aaf1

knzo25 commented on perception/lidar_transfusion/lib/preprocess/preprocess_kernel.cu:
```diff
@@ -0,0 +1,258 @@
+  unsigned char z_datatype, unsigned char intensity_datatype, unsigned char x_size,
+  unsigned char y_size, unsigned char z_size, unsigned char intensity_size, unsigned int point_step,
+  bool is_bigendian, unsigned int cloud_capacity, unsigned int points_agg, unsigned int points_size,
+  float time_lag, float * affine_past2current, float * points)
```
In this context, the transformation does not have the meaning of `affine_past2current` so it makes more sense to have a generic name

amadeuszsz replied:
https://github.com/autowarefoundation/autoware.universe/pull/6890/commits/3c0187f3d1bc8aacc640ea313016cff8b2e7aaf1

knzo25 replied:
Sorry If I was not clear enough. The change from `affine_past2current` to `affine_transform` was only in the kernel, where the nuance of transforming between time stamps was lost. :bow:

amadeuszsz replied:
Sorry for misunderstanding
https://github.com/autowarefoundation/autoware.universe/pull/6890/commits/8b59a37a133d1d1bdc7b4ae2c606415b60e0f16f

knzo25 commented on perception/lidar_transfusion/lib/preprocess/preprocess_kernel.cu:
```diff
@@ -0,0 +1,258 @@
+  float z = cast_field(z_raw, z_datatype);
+  float intensity = cast_field(intensity_raw, intensity_datatype);
+
+  float x_affined = affine_past2current[0] * x + affine_past2current[1] * y +
```
Usually (from what I have seen), the standard is `x_transformed`

amadeuszsz replied:
https://github.com/autowarefoundation/autoware.universe/pull/6890/commits/2eca809aa527823b66de2241382b14b45284a770

knzo25 commented:
A while ago (how time flies), it was decided that we would migrate towards this way of using parameters:
https://github.com/orgs/autowarefoundation/discussions/3371

While not all the nodes have implemented this (in particular centerpoint), new nodes should comply with it. Can you implement this? (there are several old PRs  if you need examples)

knzo25 commented on perception/lidar_transfusion/src/lidar_transfusion_node.cpp:
```diff
@@ -0,0 +1,161 @@
+    "~/output/objects", rclcpp::QoS(1));
+
+  published_time_pub_ = std::make_unique<tier4_autoware_utils::PublishedTimePublisher>(this);
+
```
(The comment does not apply specifically to this line)

In centerpoint there is parameter dimensionality check for the pointcloud ranges and voxel sizes.
Is there a particular reason for not doing the same in transfusion?

```
  if (point_cloud_range.size() != 6) {
    RCLCPP_WARN_STREAM(
      rclcpp::get_logger("lidar_centerpoint"),
      "The size of point_cloud_range != 6: use the default parameters.");
  }
  if (voxel_size.size() != 3) {
    RCLCPP_WARN_STREAM(
      rclcpp::get_logger("lidar_centerpoint"),
      "The size of voxel_size != 3: use the default parameters.");
  }
```

amadeuszsz replied:
Right, this might me helpful.
I added parameter descriptor as well since all the parameters shouldn't be modified. Unfortunately, it's not possible to set limits in arrays this way (it works only for primitive types). It means at the node level it's possible to set e.g. negative `voxel_size`. However, json schema (https://github.com/autowarefoundation/autoware.universe/pull/6890/commits/ac01c4a29a362b902c99e4a1b2d18e5ff5cb2b15) acts as a validation from Autoware design point of view.
https://github.com/autowarefoundation/autoware.universe/pull/6890/commits/8434721419a0612630c3072c46838fe6af8c5c86

knzo25 commented on perception/lidar_transfusion/src/lidar_transfusion_node.cpp:
```diff
@@ -0,0 +1,161 @@
+
+  std::vector<Box3D> det_boxes3d;
+  std::unordered_map<std::string, double> proc_timing;
+  bool is_success = detector_ptr_->detect(*msg, tf_buffer_, det_boxes3d, proc_timing);
```
This was already a code smell in centerpoint, but we are adding a new parameter, making it more noticeable.
Instead of using references as arguments to return values, it is recommended to use tuples.
Then, in this file, we can use structured bindings like: `auto & [is_success, det_boxes3d, proc_timing]= ...`

amadeuszsz replied:
As we discussed, RVO / NRVO will not apply in that case and avoiding the copy with structured bindings seems not trivial here. Since the current approach is working, we will leave possible improvement for future PR.

knzo25 commented on perception/lidar_transfusion/include/lidar_transfusion/transfusion_trt.hpp:
```diff
@@ -0,0 +1,114 @@
+  std::unique_ptr<NetworkTRT> network_trt_ptr_{nullptr};
+  std::unique_ptr<VoxelGenerator> vg_ptr_{nullptr};
+  std::unique_ptr<tier4_autoware_utils::StopWatch<std::chrono::milliseconds>> stop_watch_ptr_;
+  std::shared_ptr<PreprocessCuda> pre_ptr_;
```
any reason these should not be unique ptr?

knzo25 replied:
also uniform initialization to nullptr

amadeuszsz replied:
https://github.com/autowarefoundation/autoware.universe/pull/6890/commits/48200d207d459c01f9fb7dc38b33b355bc1437f0

knzo25 commented on perception/lidar_transfusion/lib/transfusion_trt.cpp:
```diff
@@ -0,0 +1,262 @@
+  voxel_num_d_ = cuda::make_unique<unsigned int[]>(voxel_num_size_);
+  voxel_idxs_d_ = cuda::make_unique<unsigned int[]>(voxel_idxs_size_);
+  points_d_ = cuda::make_unique<float[]>(config_.cloud_capacity_ * config_.num_point_feature_size_);
+  pre_ptr_.reset(new PreprocessCuda(config_, stream_));
```
In the header file I commented that these may be unique ptrs
In addition, please avoid using new and instead use std::make_...

amadeuszsz replied:
https://github.com/autowarefoundation/autoware.universe/pull/6890/commits/48200d207d459c01f9fb7dc38b33b355bc1437f0

knzo25 commented on perception/lidar_transfusion/include/lidar_transfusion/transfusion_config.hpp:
```diff
@@ -0,0 +1,158 @@
+  }
+
+  ///// INPUT PARAMETERS /////
+  std::size_t cloud_capacity_ = 1000000;
```
Please use uniform initialization
Constants should be declared  as such

amadeuszsz replied:
https://github.com/autowarefoundation/autoware.universe/pull/6890/commits/6ef95d3b6d013cc75c0184c537aefaba13543b44

knzo25 commented on perception/lidar_transfusion/lib/transfusion_trt.cpp:
```diff
@@ -0,0 +1,262 @@
+void TransfusionTRT::initPtr()
+{
+  // point cloud to voxels
+  voxel_features_size_ = config_.max_voxels_ * config_.max_num_points_per_pillar_ * 5;
```
Please do not use magic numbers. You should reuse the variables from the config header

knzo25 commented on perception/lidar_transfusion/lib/transfusion_trt.cpp:
```diff
@@ -0,0 +1,262 @@
+  proc_timing.emplace(
+    "debug/processing_time/inference_ms", stop_watch_ptr_->toc("processing/inner", true));
+
+  if (false) {
```
Why is this if(false) needed here?
I am aware that if(true) and the like are commonly used in macros, but I don't think this is the case

amadeuszsz replied:
Somehow this trick get rid of VS Code warnings for me before. However, now it is not necessary as I noticed.
https://github.com/autowarefoundation/autoware.universe/pull/6890/commits/dca8b5eaa84a99beaf7b53e9acb6a34a1596ab53

knzo25 commented on perception/lidar_transfusion/lib/transfusion_trt.cpp:
```diff
@@ -0,0 +1,262 @@
+  pre_ptr_->generateVoxels(
+    points_d_.get(), count, params_input_d_.get(), voxel_features_d_.get(), voxel_num_d_.get(),
+    voxel_idxs_d_.get());
+  unsigned int params_input_cpu;
```
The nomenclature is usually host/device, rather than cpu/gpu

amadeuszsz replied:
https://github.com/autowarefoundation/autoware.universe/pull/6890/commits/389c6cbfc28ff9abf4be285029f6a19a9101ed1b

knzo25 commented on perception/lidar_transfusion/lib/transfusion_trt.cpp:
```diff
@@ -0,0 +1,262 @@
+
+  const auto count = vg_ptr_->generateSweepPoints(msg, points_d_);
+  RCLCPP_DEBUG_STREAM(rclcpp::get_logger("lidar_transfusion"), "Generated sweep points: " << count);
+  CHECK_CUDA_ERROR(cudaStreamSynchronize(stream_));
```
Maybe I am understanding it wrong, but are there not some redundant and probably sub-optimal synchronizes around here?

amadeuszsz replied:
Yeah, that one is unnecessary since `generateSweepPoints` do the sync at the end. According to your next comment, my strategy was to call sync only if next operation was related to the device. I did some tests before and basically it does not affect performance at all. The only difference was around partial time performance measurements. The total time was same, but if I omitted some sync, TensorRT handle it itself within the next kernel. Then instead of getting e.g. `preprocess=1 ms` & `inference=1 ms` you can see `preprocess=0.01 ms` & `inference=1.99 ms`. However, I feel more safe with calling sync when looks it needed.

amadeuszsz replied:
https://github.com/autowarefoundation/autoware.universe/pull/6890/commits/fe15baf01fd4f96a4d2c53a6abd6e49468cd68e0

knzo25 commented:
@amadeuszsz 
Reading the preprocessing -> inference -> postprocessing part, I see several syncs that are needed to check the error code and abort midway when needed. Are all of these necessary? As a reference, centerpoint does not. Would the speedup product of removing these be worth it?

(I know that the one related the the optimization plan is important unless we have min voxels = 1)

knzo25 commented on perception/lidar_transfusion/lib/transfusion_trt.cpp:
```diff
@@ -0,0 +1,262 @@
+      std::exp(box_output_cpu[i + 3 * config_.num_proposals_]),
+      std::exp(box_output_cpu[i + 4 * config_.num_proposals_]),
+      std::exp(box_output_cpu[i + 5 * config_.num_proposals_]),
+      std::atan2(dir_cls_output_cpu[i], dir_cls_output_cpu[i + config_.num_proposals_])});
```
In centerpoint, we discovered that the values from the cos/sin estimation do not always have a unit norm (why would they anyways, ML does not assure that).
There are cases where the objects have orientation error of 90 degrees, that causes several overrides. We found out that these cases are highly correlated to the cases where cos/sin values are both close to 0 (or have a norm quite different to 1). I implemented a simple filter for that ages ago for centerpoint. Can you evaluate if that phenomena happens in transfusion? If so, can you reimplement that functionality in transfusion?

(I commented this in the cpu implementation rather than the cuda one due to the review order, but it also applies there as well)

amadeuszsz replied:
https://github.com/autowarefoundation/autoware.universe/pull/6890/commits/7a054cca63abac2ac20990cfff676c22ff92e6cc
Right, I forgot to check it for host processing. Moreover, I checked the output with  `yaw_norm < config_.yaw_norm_thresholds_[class_id]` condition and some frames generate detection where `max_score >= config_.score_threshold_`.  These detection had 90 degrees orientation error. Therefore we need that filter for `lidar_transfusion` as well.

knzo25 commented on perception/lidar_transfusion/lib/preprocess/voxel_generator.cpp:
```diff
@@ -0,0 +1,156 @@
+  cloud_data_d_ = cuda::make_unique<unsigned char[]>(config_.cloud_capacity_ * MAX_CLOUD_STEP_SIZE);
+  points_cpu_.resize(config_.cloud_capacity_ * config_.num_point_feature_size_);
+  affine_past2current_d_ = cuda::make_unique<float[]>(AFF_MAT_SIZE);
+  pre_ptr_.reset(new PreprocessCuda(config_, stream_));
```
Please avoid using new and use unique ptr when possible

amadeuszsz replied:
https://github.com/autowarefoundation/autoware.universe/pull/6890/commits/48200d207d459c01f9fb7dc38b33b355bc1437f0

knzo25 commented on perception/lidar_transfusion/include/lidar_transfusion/preprocess/voxel_generator.hpp:
```diff
@@ -0,0 +1,70 @@
+
+#ifndef LIDAR_TRANSFUSION__PREPROCESS__VOXEL_GENERATOR_HPP_
+#define LIDAR_TRANSFUSION__PREPROCESS__VOXEL_GENERATOR_HPP_
+#define AFF_MAT_SIZE 16         // 4x4 matrix
```
Please consider using constexpr (they have type safety)

amadeuszsz replied:
https://github.com/autowarefoundation/autoware.universe/pull/6890/commits/6100a0548c41d636f5d49714fe0b4b19963bdcc2

knzo25 commented on perception/lidar_transfusion/lib/preprocess/preprocess_kernel.cu:
```diff
@@ -0,0 +1,258 @@
+  CHECK_CUDA_ERROR(cudaStreamSynchronize(stream_));
+}
+
+__global__ void generateVoxels_random_kernel(
```
nit: usually if the call order is A call B, B is first than A in the header
(nits can be ignored and "resolved" by you)

amadeuszsz replied:
With your proposed order, compiler can't see function definition because the function was not declared.

knzo25 commented on perception/lidar_transfusion/include/lidar_transfusion/preprocess/preprocess_kernel.hpp:
```diff
@@ -0,0 +1,72 @@
+namespace lidar_transfusion
+{
+
+class PreprocessCuda
```
I do not think that this header is the right location for this class (it is meant for kernels). It probably belongs in its own header

amadeuszsz replied:
This class follows same approach as `postprocess_kernel.hpp`, imo it's right location or I don't see your point of view. If so, please extend your comment

amadeuszsz replied:
Imo if there is a reason and general principle of not using OOP in .cu files, let me know please. In that case I would rather refactor this code and `lidar_centerpoint` as well

knzo25 commented on perception/lidar_transfusion/lib/preprocess/preprocess_kernel.cu:
```diff
@@ -0,0 +1,258 @@
+  float x = points[point_idx * 5];
+  float y = points[point_idx * 5 + 1];
+  float z = points[point_idx * 5 + 2];
+  float w = points[point_idx * 5 + 3];
```
nit: I do not think calling intensity as w is semantically adequate.
(the naming is probably a leftover from the float4 from centerpoint, which case there is no avoiding it)

amadeuszsz replied:
https://github.com/autowarefoundation/autoware.universe/pull/6890/commits/63a30098d1c9f6e71e50e96ab0d46acb811a3c05

knzo25 commented on perception/lidar_transfusion/lib/network/network_trt.cpp:
```diff
@@ -0,0 +1,231 @@
+
+bool NetworkTRT::validateNetworkIO()
+{
+  if (engine->getNbIOTensors() != NetworkIO::ENUM_SIZE) {
```
Thank your creating some checks (I think we talked about this a while ago).
Although the number of I/O is a good one, it may also be useful to check the dimensionality of some of the tensors.
For example, in centerpoint there is:

```
  for (std::size_t ci = 0; ci < out_channel_sizes_.size(); ci++) {
    auto out_name = network.getOutput(ci)->getName();

    if (
      out_name == std::string("heatmap") &&
      network.getOutput(ci)->getDimensions().d[1] != static_cast<int32_t>(out_channel_sizes_[ci])) {
      tensorrt_common::LOG_ERROR(logger_)
        << "Expected and actual number of classes do not match" << std::endl;
      return false;
    }
    auto out_dims = nvinfer1::Dims4(
      config_.batch_size_, out_channel_sizes_[ci], config_.down_grid_size_y_,
      config_.down_grid_size_x_);
    profile->setDimensions(out_name, nvinfer1::OptProfileSelector::kMIN, out_dims);
    profile->setDimensions(out_name, nvinfer1::OptProfileSelector::kOPT, out_dims);
    profile->setDimensions(out_name, nvinfer1::OptProfileSelector::kMAX, out_dims);
  }
```

amadeuszsz replied:
https://github.com/autowarefoundation/autoware.universe/pull/6890/commits/e83d0bffa1142767c7b968b4661a140390f2e34a
Agree.
First, the optimization profile has to match that one in TensorRT engine. It prevents a scenario when user changed the config but forgot to delete previous engine file.
Secondly the tensors shapes have to match actual network. The `lidar_transfusion` handles inference for specific network configuration. Some parameters are exposed, but most of them are defined in header file. It prevents a scenario when user provided custom network model which may not follow our assumptions (e.g. num of classes).

knzo25 replied:
Thanks ! :star_struck:

knzo25 commented on perception/lidar_transfusion/include/lidar_transfusion/visibility_control.hpp:
```diff
@@ -0,0 +1,37 @@
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#ifndef LIDAR_TRANSFUSION__VISIBILITY_CONTROL_HPP_
```
Is there any reason for this file?
I don't think we support anything other than linux

amadeuszsz replied:
My motivation was fact that some of Autoware packages already consider `visibility_control`. The Coding guidelines in Autoware documentation section do not cover it yet, so I assume the other way: packages may missing this header. If this header is somehow a problem for us, let me know

knzo25 replied:
Wow, I  was not aware of the details
https://gcc.gnu.org/wiki/Visibility

Let's definitively keep this :+1:

knzo25 commented on perception/lidar_transfusion/include/lidar_transfusion/lidar_transfusion_node.hpp:
```diff
@@ -0,0 +1,71 @@
+{
+public:
+  explicit LidarTransfusionNode(const rclcpp::NodeOptions & options);
+  void cloudCallback(const sensor_msgs::msg::PointCloud2::ConstSharedPtr msg);
```
nit: the cloudCallback should be private or protected

amadeuszsz replied:
https://github.com/autowarefoundation/autoware.universe/pull/6890/commits/46508831d2f342a5dec6d8b4894790771d9e4809

knzo25 commented on perception/lidar_transfusion/include/lidar_transfusion/lidar_transfusion_node.hpp:
```diff
@@ -0,0 +1,71 @@
+
+  NonMaximumSuppression iou_bev_nms_;
+
+  std::unique_ptr<TransfusionTRT> detector_ptr_;
```
nit: pointers should use uniform initialization with a nullptr

amadeuszsz replied:
https://github.com/autowarefoundation/autoware.universe/pull/6890/commits/be8914182ee63fc1962e68927fb11b7e88c85b1c

knzo25 commented on perception/lidar_transfusion/include/lidar_transfusion/transfusion_trt.hpp:
```diff
@@ -0,0 +1,114 @@
+
+  std::unique_ptr<NetworkTRT> network_trt_ptr_{nullptr};
+  std::unique_ptr<VoxelGenerator> vg_ptr_{nullptr};
+  std::unique_ptr<tier4_autoware_utils::StopWatch<std::chrono::milliseconds>> stop_watch_ptr_;
```
uniform initialization to nullptr

amadeuszsz replied:
https://github.com/autowarefoundation/autoware.universe/pull/6890/commits/48200d207d459c01f9fb7dc38b33b355bc1437f0

knzo25 commented on perception/lidar_transfusion/include/lidar_transfusion/transfusion_trt.hpp:
```diff
@@ -0,0 +1,114 @@
+
+  // input of pre-process
+
+  unsigned int voxel_features_size_ = 0;
```
uniform initialization

amadeuszsz replied:
https://github.com/autowarefoundation/autoware.universe/pull/6890/commits/48200d207d459c01f9fb7dc38b33b355bc1437f0

knzo25 reviewed (changes requested):
Great work !
Most of my comments are regarding code style/standards and optimization ideas 
:rocket:

knzo25 commented:
@amadeuszsz 
The person who "resolves" the comments is the one who writes them. When they are addressed, please paste the commit in which it was done. If there are several changes, it is appreciated if you split them in different commits

knzo25 commented:
@amadeuszsz 
A comment just in case for future reference. One of the good points of uniform intialization is that it is guaranteed to assign the correct zero to the primitive types and nullptr for pointers when using empty curly braces {}

amadeuszsz commented:
@knzo25 
> While doing other task that was asked of me, I was looking at the whole perception task and saw the preprocessing being done in cpu in centerpoint. I kind of forgot you told me you were doing it already in gpu for transfusion so I ended up doing double work and implemented it myself there as well 🙇
> 
> When I sent the PR I remembered, and after comparing the implementation, while quite similar, they have different overall costs, so I wonder if you can change the implementation here so it has around the same number of operations:
> 
> Let `K` be the total number of frames used for inference and `N` the number of points in a pointcloud. Assuming that the queue if full, the preprocessing before proper voxelization cost would be as follows:
> 
> transfusion: enqueue:
> 
> * N* copy(host->host)
>   transform and concat:
> * move to gpu: K_N_copy(host->device)
> * kernel: K_N_transform(device->device)
> 
> (proposed) centerpoint: enqueue:
> 
> * N * copy (host->device)
>   transform and concat:
> * kernel: K_N_kernel(device->device)

I modified the code as you suggested. The preprocessing decreased from 2.15 ± 0.55 to 1.21 ± 0.25 [ms] with two clouds in cache (performance differs from PR's init benchmark due to different load). The breaking changes make harder to keep host processing (you predict that before 😃), therefore I just removed them.

> After reading your kernels seriously this time, I realized that you went to great lengths to make the operations general. Usually, when writing kernels (cpu or gpu), you want to avoid all the not strictly needed branches and loops. Could you specialize the kernels so that the fastest implementation is used on runtime? (maybe leaving your generic version to the case that no fast implementation is available)
> 
> Also, since we control the driver, you usually would not need to pay much attention to reverting the endianness

Changed. Now the package do not support other point cloud formats. However, there is init validation and we throw exception with appropriate logs.

https://github.com/autowarefoundation/autoware.universe/pull/6890/commits/98e31a99191553352ebc8f94a372f89e7ae348f8
Seems it's last requested modification. Since it was a big change, apart of code review please test it again during the runtime.

knzo25 commented on perception/lidar_transfusion/lib/preprocess/pointcloud_densification.cpp:
```diff
@@ -90,7 +92,19 @@ void PointCloudDensification::enqueue(
+  auto points_d = cuda::make_unique<float[]>(
+    sizeof(uint8_t) * msg.width * msg.height * msg.point_step / sizeof(float));
+
+  CHECK_CUDA_ERROR(cudaMemcpyAsync(
```
This is a question rather than a review comment that requires changes from your part.
I had understood that async did not really work if the memory was not pinned? Did I misunderstand that part?

knzo25 replied:
@amadeuszsz 
Talking with Hirabayashi-san, we kind of got the impression that the operation is not fully parallelized (pageable to pinned may not be async and then the pinned to device is async). If you have any further information, please tell us, otherwise please just close the comment

amadeuszsz replied:
> pageable to pinned may not be async and then the pinned to device is async

Right, but docs are kind of unclear - how to interpret "_may_" in "_the driver may synchronize with the stream_" [here](https://docs.nvidia.com/cuda/cuda-runtime-api/api-sync-behavior.html#api-sync-behavior__memcpy-async)? There was an update between [12.0.0](https://docs.nvidia.com/cuda/archive/12.0.0/cuda-runtime-api/api-sync-behavior.html#api-sync-behavior__memcpy-async) and [12.1.0](https://docs.nvidia.com/cuda/archive/12.1.0/cuda-runtime-api/api-sync-behavior.html#api-sync-behavior__memcpy-async) documentation (we target 12.2.0), the old one is unambiguous and I'm not sure if it's only text refactor or there was some changes under the hood. Anyway, the API call type (e.g. `cudaMemcpyAsync` used for sync opertation) is not that important since CUDA will force sync whenever it's needed. For clarity, I added stream synchronization each time we access the device memory. I didn't observe any overhead and it will return error if the memory was corrupted in previous operation -> if something wrong happen, users will come with appropriate logs.

knzo25 reviewed (changes requested):
@amadeuszsz 
Thank you for handling all the comments. before merging the PR though, can you turn back the default model to centerpoint?

knzo25 reviewed (approved):
LGTM !
:rocket:

knzo25 commented:
@YoshiRi 
Due to having touched the launchers, we also need a review from one of the maintainers of that package.
Since it is only a few lines, could you make a short  review of that part?

YoshiRi reviewed (approved):
LGTM for launcher part

knzo25 commented on perception/lidar_transfusion/README.md:
```diff
@@ -0,0 +1,128 @@
+
+| Name                             | Type         | Default Value | Description                                                                                        |
+| -------------------------------- | ------------ | ------------- | -------------------------------------------------------------------------------------------------- |
+| `class_names`                    | list[string] | -             | Class names for 3D object detection.                                                               |
```
@amadeuszsz 
Sorry for the late comment, but since you added the schema file, could you change the README.md so it uses the schema instead?

amadeuszsz replied:
Done 👍🏻
